pkg/util/fxutil/fxoptional/component.go

```diff
@@ -0,0 +1,97 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package fxoptional
+
+import "github.com/DataDog/datadog-agent/pkg/util/optional"
+
+// Component is an optional component type.
+// The main purpose of this component is to make it explicit that the component may not be present.
+// Here is an full example of how to use it:
+//
+//	// ------------------------------------------
+//	// Component.go
+//	// ------------------------------------------
+//	type Component fxoptional.Component[MyInterface]
+//
+//	type MyInterface interface {
+//	}
+//
+//	// ------------------------------------------
+//	// Another file implements the component
+//	// ------------------------------------------
+//
+//	// Module defines the fx options for this component.
+//	func Module() fxutil.Module {
+//		return fxutil.Component(
+//			fx.Provide(newMyComponent))
+//	}
+//
+//	type dependencies struct {
+//		fx.In
+//
+//		Config    config.Component
+//		Lifecycle fx.Lifecycle
+//	}
+//
+//	type myComponent struct{}
+//
+//	func newMyComponent(deps dependencies) Component {
+//		if !deps.Config.GetBool("mycomponent.enabled") {
+//			return fxoptional.None[MyInterface]()
+//		}
+//
+//		// Use lifecycle hooks if needed by the component
+//		deps.Lifecycle.Append(fx.Hook{
+//			OnStart: func(context.Context) error {
+//				return nil
+//			},
+//			OnStop: func(context.Context) error {
+//				return nil
+//			},
+//		})
+//		return fxoptional.New[MyInterface](myComponent{})
+//	}
+//
+// // ------------------------------------------
+// // Usage
+// // ------------------------------------------
+//
+//	func UseMyComponent(comp Component) {
+//		myComp, ok := comp.Get()
+//		if ok {
+//			// Do something with myComp
+//		}
+//	}
+//
+// It is recommended to use fxoptional.Component instead of optional.Option[Component].
+// Consider the following example of code
+//
+//	 func run(myComponent Component) {
+//	 }
+//
+//	func MyFunc() {
+//		fxutil.OneShot(run, Module())
+//	}
+//
+// This code fails when using optional.Option[Component] because the type of the component is not Component but optional.Option[Component].
+// The user must know that Module() provides an way to create an optional.Option[Component] and not a Component.
+// This is not the case with fxoptional.Component.
+// Note: This issue already happened in the past and was hard to debug.
+type Component[T any] interface {
+	Get() (T, bool)
```

Comment:
Does it only work for components whose interface is just `Get()`? 

Sorry if the question is naive, I'm not very familiar with Go generics 😄

Reply:
I see 😄 

Thank you so much for the explanation. I was mostly confused because the forwarder component interface also used the `Get()` function.

```go
type Component interface {
	// Get the forwarder instance if it exists.
	Get() (defaultforwarder.Forwarder, bool)
}
```

Reply:
The following might not be super accurate, but I just meant that the type you really care about (in the case the actual forwarder with all its logic and state) is the `T`.
> It is not an fx component itself, the component is the embedded T.

It was the same thing with the forwarder Component, it wasn't the actual forwarder but an `optional` wrapper around it.